### PR TITLE
DI-920 Changes for Medicover internal AF VCF

### PR DESCRIPTION
**Pull request by rklocke**
- Add run mode choices where in mode 'no_qc' we can find Medicover VCFs and not find the QC status files in the GRCh37 project like we would for CEN or TWE
- Small changes to bash script to give better logging and check memory usage etc, use fields 2 and 3 of input file (since we longer output the pandas index from the Python file) and name output file by input file
- Update README for changes

- - -
This change is [Reviewable](https://reviewable.io/reviews/eastgenomics/RD_requests/19)

## Summary by CodeRabbit

- **New Features**
	- Introduced a new command-line argument `--run_mode` for the VCF merging script, allowing users to choose between `find_qc` and `no_qc` modes.
	- Added functionality to locate and process VCF files directly.
	- Enhanced logging and resource usage reporting in the merging script.

- **Documentation**
	- Updated README to clarify operational modes and provide example commands, improving user understanding of the scripts.

- **Bug Fixes**
	- Improved error handling and output messaging for project handling in the VCF processing script.

- **Chores**
	- Updated output file naming conventions for clarity and consistency.

### CODE REVIEW

**Review by RSWilson1 — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/RD_requests/19)* status: all files reviewed, 14 unresolved discussions (waiting on @coderabbitai[bot] and @rklocke)

___
*[`DI-435/find_vcfs_to_merge.py` line 357 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OCdKSEn13I9suQnNSs0:-OCdKSEn13I9suQnNSs1:b9kkwv4) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/find_vcfs_to_merge.py#L357)):*
> ```Python
> def find_medicover_vcf_files(projects):
>     """
>     Find VCF files in a project
> ```

maybe add some more context to this docstring
___
*[`DI-435/find_vcfs_to_merge.py` line 369 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OCdCzMq8Wt1XhSavih2:-OCdCzMq8Wt1XhSavih3:b1yzho7) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/find_vcfs_to_merge.py#L369)):*
> ```Python
>         list of dicts, each representing a VCF file in DX
>     """
>     sub_remove = {
> ```

this naming could be clearer
___
*[`DI-435/find_vcfs_to_merge.py` line 381 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OCdD9NJ25-0H1GRXVy2:-OCdD9NJ25-0H1GRXVy3:b-pgl6jj) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/find_vcfs_to_merge.py#L381)):*
> ```Python
>         for vcf_file in vcf_files:
>             file_id = vcf_file["describe"]["id"]
>             file_name = vcf_file["describe"]["name"]
> ```

maybe a comment to explain the logic?
___
*[`DI-435/merge_VCF_AF.sh` line 94 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OCdCTxh9EGF2_Zg2f7d:-OCdCTxh9EGF2_Zg2f7e:bp3s396) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/merge_VCF_AF.sh#L94)):*
> ```Shell
> mkdir -p norm
> echo "Normalising VCFs"
> find . -maxdepth 1 -name '*.vcf.gz' -print0 | xargs -0 -P "${num_proc}" -I{} bcftools norm -m -any -f "${genome}" -Oz "{}" -o norm/"$(basename {})"
> ```

NICE! I was trying to get this to work but kept getting odd errors.
___
*[`DI-435/merge_VCF_AF.sh` line 134 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OChQVYS2xg4p7ibhdOA:-OChQVYS2xg4p7ibhdOB:bqu334f) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/merge_VCF_AF.sh#L134)):*
> ```Shell
> dx upload "final_merged_${file_prefix}.vcf.gz"
> dx upload "final_merged_${file_prefix}.vcf.gz.tbi"
> _get_peak_usage
> ```

This is nifty :thumbsup:

File: DI-435/find_vcfs_to_merge.py [hunk 1]
```diff
@@ -48,6 +48,18 @@ def parse_args() -> argparse.Namespace:
...
+        "-r",
+        "--run_mode",
+        choices=['find_qc', 'no_qc'],
+        required=True,
+        help=(
+            "Runmode - either 'find_qc' where QC status files are "
+            "searched for and used to find failed samples, or 'no_qc'"
+            "where no QC status files are searched for"
+        )
+    )
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/find_vcfs_to_merge.py]: Done.

File: DI-435/find_vcfs_to_merge.py [hunk 2]
```diff
@@ -335,6 +352,53 @@ def get_failed_samples(qc_status_df):
...
+                    "file_id": file_id
+                }
+            )
+
+    print(
+        f"\nFound {len(all_sample_vcfs)} VCF files in GRCh38 projects"
+    )
+
+    return all_sample_vcfs
+
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/find_vcfs_to_merge.py]: I have added extra detail but for consistency I will leave the type hints out

File: DI-435/find_vcfs_to_merge.py [hunk 3]
```diff
@@ -409,84 +473,136 @@ def main():
     projects_to_print = '\n\t'.join([
         f"{x['describe']['name']} - {x['id']}" for x in b38_projects
     ])
-    print(f"\n{len(b38_projects)} projects found:\n\t{projects_to_print}")
-
-    # Get QC status files from b37 projects and read them in
-    all_qc_files, missing_projects, b38_project_subset = get_qc_files(
-        b38_projects, args.start, args.end
+    print(
+        f"\n{len(b38_projects)} GRCh38 projects found:\n\t{projects_to_print}"
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/find_vcfs_to_merge.py:481]: Not needed

File: DI-435/find_vcfs_to_merge.py [hunk 4]
```diff
@@ -335,6 +352,57 @@ def get_failed_samples(qc_status_df):
...
+    -------
+    all_sample_vcfs : list
+        list of dicts, each representing a VCF file in DX
+    """
+    # Parts of filename we don't want in the sample name
+    strings_to_remove = {
+        '_Wdh': '',
+        '_Wdh2': '',
+        '_Whd3': '',
+    }
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/find_vcfs_to_merge.py:374]: Actually not a typo

File: DI-435/merge_VCF_AF.sh [hunk 5]
```diff
@@ -12,6 +12,33 @@ genome=$2
 job="${DX_JOB_ID}"
 num_proc="$(grep -c ^processor /proc/cpuinfo)"
 
+exec 1> >(tee -a -i "/home/dnanexus/dx_stdout")
+exec 2> >(tee -a -i "/home/dnanexus/dx_stderr")
+
+# Prefixes all lines of commands written to stdout with datetime
+PS4='\000[$(date)]\011'
+export TZ=Europe/London
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/merge_VCF_AF.sh:20]: This works fine as is

File: DI-435/merge_VCF_AF.sh [hunk 6]
```diff
@@ -12,6 +12,33 @@ genome=$2
...
+
+# set frequency of instance usage in logs to 10 seconds
+sudo kill $(ps aux | grep pcp-dstat | head -n1 | awk '{print $2}')
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/merge_VCF_AF.sh:23]: I added this part as advised by Jethro and DNAnexus didn't allow kill without sudo - I'm not sure I understand enough to know if changing this is fine..

File: DI-435/merge_VCF_AF.sh [hunk 7]
```diff
@@ -12,6 +12,33 @@ genome=$2
...
+/usr/bin/dx-dstat 10
+
+_get_peak_usage() {
+    : '''
+    Reports the peak memory and storage usage from dstat, to be called at end of script
+    '''
+    dx watch $DX_JOB_ID --no-follow --quiet > job.log
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/merge_VCF_AF.sh]: Done.

File: DI-435/README.md [hunk 8]
```diff
@@ -1,43 +1,83 @@
...
-**Searching Dates**
-These dates restrict the projects collated to only GRCh38 projects
-which have corresponding GRCh37 projects which were created within the specified dates.
+Example command to find Medicover VCFs with no QC status files:
+```
+python3 find_vcfs_to_merge.py \
+    --assay "*TWE38M" \
+    --outfile_prefix TWE38M \
+    --run_mode no_qc
+```
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/README.md]: Done.

File: DI-435/README.md [hunk 9]
```diff
@@ -1,43 +1,83 @@
...
 
-Example Python script command:
-`python3 find_vcfs_to_merge.py --assay "*CEN38" --end 2024-05-03 --outfile_prefix CEN38`
+#### Inputs
+- `-a --assay`: The GRCh38 project search term for in DNAnexus, e.g. `"*CEN38"`.
+- `-o --outfile_prefix`: Prefix to use to name the output files, e.g. `CEN38`.
+- `-r --run_mode`: A choice of whether to find and use QC status files (`find_qc`) or not (`no_qc`).
+In `find_qc` mode the following inputs are allowed:
+  - `-s --start (optional)`: A date used to find DNAnexus (GRCh37) projects created after, see searching dates section below.
+  - `-e --end (optional)`: A date used to find DNAnexus (GRCh37) projects created before, see searching dates section below.
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/README.md]: Done.

File: DI-435/README.md [hunk 10]
```diff
@@ -1,43 +1,83 @@
...
+
+#### no_qc mode
+How the script works in `no_qc` mode:
+1. Finds all DNAnexus projects with suffix `--assay`
+2. Finds all raw VCFs in each of the DNAnexus projects.
+3. Removes all samples which are duplicates.
+4. Creates a final list of all VCFs to merge and writes this to file with `--outfile_prefix`.
+##### Outputs
+- A TSV listing the VCF files to merge (named by `{outfile_prefix}_files_to_merge.txt`)
+- A TSV of all duplicate files which have been removed (`{outfile_prefix}_all_dup_rows.txt`)
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/README.md]: Done.

File: DI-435/README.md [hunk 11]
```diff
@@ -1,43 +1,83 @@
...
+```
+bash merge_VCF_AF.sh \
+    CEN38_files_to_merge.txt \
+    GRCh38_GIABv3_no_alt_analysis_set_maskedGRC_decoys_MAP2K3_KMT2C_KCNJ18_noChr.fasta
+```
 
 Output:
-Merged VCF file and index named from the job ID given, i.e. `final_merged_job-Gjb39Z04bxf82XZ12gPJ2bbV.vcf.gz` and `final_merged_job-Gjb39Z04bxf82XZ12gPJ2bbV.vcf.gz.tbi` which are both uploaded to the DNAnexus project the cloud workstation is running within.
\ No newline at end of file
+Merged VCF file and index named from the input files_to_merge txt file, i.e. `final_merged_CEN38_files_to_merge.vcf.gz` and `final_merged_CEN38_files_to_merge.vcf.gz.tbi` which are both uploaded to the DNAnexus project the cloud workstation is running within.
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/README.md]: Done.

File: DI-435/README.md [hunk 12]
```diff
@@ -1,43 +1,84 @@
...
-**Searching Dates**
-These dates restrict the projects collated to only GRCh38 projects
-which have corresponding GRCh37 projects which were created within the specified dates.
+Example command to find Medicover VCFs with no QC status files:
+```bash
+python3 find_vcfs_to_merge.py \
+    --assay "*TWE38M" \
+    --outfile_prefix TWE38M \
+    --run_mode no_qc
+```
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/README.md:33]: Already clarified that the inputs are only allowed in find_qc mode

File: DI-435/README.md [hunk 13]
```diff
@@ -1,43 +1,84 @@
...
-- A TSV listing the VCF files for all non-validation samples to merge (named by `{outfile_prefix}_files_to_merge.txt`)
-- A CSV of all validation samples found (`{outfile_prefix}_validation_samples.csv`)
-- A CSV of all projects within search but missing QC file in DNAnexus and therefore not included (`{outfile_prefix}_projects_missing_QC.csv`).
+- `-a --assay`: The GRCh38 project search term for in DNAnexus, e.g. `"*CEN38"`.
+- `-o --outfile_prefix`: Prefix to use to name the output files, e.g. `CEN38`.
+- `-r --run_mode`: A choice of whether to find and use QC status files (`find_qc`) or not (`no_qc`).
+
+In `find_qc` mode the following inputs are allowed:
+  - `-s --start (optional)`: A date used to find DNAnexus (GRCh37) projects created after, see searching dates section below.
+  - `-e --end (optional)`: A date used to find DNAnexus (GRCh37) projects created before, see searching dates section below.
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/README.md:43]: It's clear already

File: DI-435/README.md [hunk 14]
```diff
@@ -1,43 +1,84 @@
...
+```bash
+bash merge_VCF_AF.sh \
+    CEN38_files_to_merge.txt \
+    GRCh38_GIABv3_no_alt_analysis_set_maskedGRC_decoys_MAP2K3_KMT2C_KCNJ18_noChr.fasta
+```
 
 Output:
-Merged VCF file and index named from the job ID given, i.e. `final_merged_job-Gjb39Z04bxf82XZ12gPJ2bbV.vcf.gz` and `final_merged_job-Gjb39Z04bxf82XZ12gPJ2bbV.vcf.gz.tbi` which are both uploaded to the DNAnexus project the cloud workstation is running within.
\ No newline at end of file
+Merged VCF file and index named from the input files_to_merge txt file, i.e. `final_merged_CEN38_files_to_merge.vcf.gz` and `final_merged_CEN38_files_to_merge.vcf.gz.tbi` which are both uploaded to the DNAnexus project the cloud workstation is running within.
```
[the remark this answers is not in the document]
Comment (rklocke) [reply DI-435/README.md:84]: Not required

**Review by rklocke — COMMENTED** [review]
:lgtm:

> Dismissed @coderabbitai[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-:-OCrKmS2ElAr-qqxPFbe:bnfp4nl)* status: 0 of 3 files reviewed, 19 unresolved discussions (waiting on @coderabbitai[bot] and @RSWilson1)

___
*[`DI-435/find_vcfs_to_merge.py` line 357 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OCdKSEn13I9suQnNSs0:-OCrISXw2E5oRvvVhidz:b-896fix) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/find_vcfs_to_merge.py#L357)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

maybe add some more context to this docstring
</blockquote></details>

Done.
___
*[`DI-435/find_vcfs_to_merge.py` line 369 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OCdCzMq8Wt1XhSavih2:-OCrITgJ09EZWS1mCAcG:b-896fix) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/find_vcfs_to_merge.py#L369)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

this naming could be clearer
</blockquote></details>

Done.
___
*[`DI-435/find_vcfs_to_merge.py` line 381 at r1](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-OCdD9NJ25-0H1GRXVy2:-OCrIUUG4cv2zL7P0Wl1:b-896fix) ([raw file](https://github.com/eastgenomics/RD_requests/blob/2a8721ece7f2c8408e94ffbeebffca660317c3f4/DI-435/find_vcfs_to_merge.py#L381)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

maybe a comment to explain the logic?
</blockquote></details>

Done.

File: DI-435/merge_VCF_AF.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (RSWilson1) [reply DI-435/merge_VCF_AF.sh:23]: yeah that's fine then

**Review by RSWilson1 — APPROVED** [review]
Looks good, dismissed some coderabbit comments as not needed.

> Reviewed 3 of 3 files at r2, all commit messages.
> Dismissed @coderabbitai[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-:-OCrcWCt7Qynjjgz2Cmi:bu5sskk)* status: all files reviewed, 15 unresolved discussions (waiting on @rklocke)

**Review by RSWilson1 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/eastgenomics/RD_requests/19#-:-OCrcf7wET-RvoR1V4tZ:bnfp4nl)* status: all files reviewed, 15 unresolved discussions (waiting on @rklocke)

**Review by RSWilson1 — APPROVED** [review]
> Dismissed @coderabbitai[bot] from 14 discussions.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/RD_requests/19)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @rklocke)

**Pull request merged**